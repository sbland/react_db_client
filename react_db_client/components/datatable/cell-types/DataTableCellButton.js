import React, { useEffect, useMemo, useState } from 'react';
import PropTypes from 'prop-types';

export const DataTableCellButton = ({
  columnData: { action, uid, btnLabel },
  rowData,
  rowId,
  cellData,
  editMode,
  focused,
  resetValue, // called to end edit mode
}) => {
  const actionFunc = useMemo(() => {
    const a = action
      ? () => action(rowId, cellData, rowData)
      : () => console.warn('Button has no action');
    return () => {
      resetValue();
      a();
    };
  }, [rowId, cellData, action, rowData, resetValue]);

  useEffect(() => {
    if (editMode && focused) {
      actionFunc();
      resetValue();
    }
  }, [editMode, focused, actionFunc, resetValue]);

  const className = [
    'dataTableCellData',
    'dataTableCellData-button',
    `${uid}_cellBtn`,
  ].join(' ')

  const cellDataLabel = btnLabel ||
    cellData && typeof cellData === 'object' ? cellData.label || cellData.name : cellData;
  return (
    <div className={className}>
      <button type="button" onClick={() => {}}>
        {/* CLick disabled as button press handled by focus and edit mode */}
        {/* <button type="button" onClick={() => setCallAction(true)}> */}
        {cellDataLabel}
      </button>
    </div>
  );
};

DataTableCellButton.propTypes = {
  rowId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  cellData: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.shape({ name: PropTypes.string, label: PropTypes.string }),
  ]),
  columnData: PropTypes.shape({
    action: PropTypes.func,
  }).isRequired,
  rowData: PropTypes.shape({}).isRequired,
  focused: PropTypes.bool.isRequired,
  editMode: PropTypes.bool.isRequired,
  resetValue: PropTypes.func.isRequired,
};
DataTableCellButton.defaultProps = {
  cellData: '',
};
