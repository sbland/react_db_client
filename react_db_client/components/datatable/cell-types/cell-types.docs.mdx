---
description: 'A CellTypes component.'
labels: ['datatable']
---
