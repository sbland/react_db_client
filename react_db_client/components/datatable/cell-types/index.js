export { DataTableCellButton } from './DataTableCellButton';
export { DataTableCellEntity } from './DataTableCellEntity';
export { DataTableCellLink } from './DataTableCellLink';
export { DataTableCellNumber } from './DataTableCellNumber';
export { DataTableCellPopup } from './DataTableCellPopup';
export { DataTableCellReadOnly } from './DataTableCellReadOnly';
export { DataTableCellSelect } from './DataTableCellSelect';
export { DataTableCellText } from './DataTableCellText';
export { DataTableCellToggle } from './DataTableCellToggle';

export { defaultComponentMap } from './default-component-map';
