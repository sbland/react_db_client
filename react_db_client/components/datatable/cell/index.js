export { Cell } from './cell';
export { DataTableCellHoverWrap } from './cell-wrappers';
