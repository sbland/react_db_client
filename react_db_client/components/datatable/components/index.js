// export { DataTableCellHoverWrap } from './DataTableCell';
export { DataTableTotals } from './DataTableTotals';
export { DataTableHeadings } from './DataTableHeadings';
