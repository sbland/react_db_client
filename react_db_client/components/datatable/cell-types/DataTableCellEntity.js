import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { PopupPanel } from '@react_db_client/components.popup-panel';
import { SearchAndSelect } from '@react_db_client/components.search-and-select';
import { filterTypes } from '@react_db_client/constants.client-types';
import { DefaultCellInnerStyle } from './style';
// TODO: Make sure apiGetDocuments is passes to this component
// import { apiGetDocuments } from '../../../Api/Api';

export const SelectEntityPanel = ({
  apiGetDocuments,
  currentSelection,
  collection,
  headings,
  handleSubmit,
}) => (
  <DefaultCellInnerStyle className="entitySelector">
    <SearchAndSelect
      searchFunction={(filters) => apiGetDocuments(collection, filters)}
      initialFilters={[
        {
          uid: 'searchphrase',
          field: 'searchPhrase',
          operator: 'contains',
          value: currentSelection || '',
        },
      ]}
      availableFilters={{
        searchPhrase: {
          uid: 'searchPhrase',
          label: 'Search Phrase',
          type: 'text',
        },
        ...headings,
      }}
      handleSelect={handleSubmit}
      headings={
        headings || [
          {
            uid: 'uid',
            label: 'UID',
            type: filterTypes.text,
          },
          {
            uid: 'type',
            label: 'Type',
            type: filterTypes.text,
          },
          {
            uid: 'name',
            label: 'Name',
            type: filterTypes.text,
          },
        ]
      }
    />
  </DefaultCellInnerStyle>
);

SelectEntityPanel.propTypes = {
  currentSelection: PropTypes.string.isRequired,
  collection: PropTypes.string.isRequired,
  headings: PropTypes.shape({
    uid: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    type: PropTypes.oneOf(Object.keys(filterTypes)).isRequired,
    value: PropTypes.any,
  }).isRequired,
  handleSubmit: PropTypes.func.isRequired,
};

/**
 * Data Cell Entity
 *
 * @param {*} {
 *   cellData,
 *   updateData,
 * }
 * @returns
 */
export const DataTableCellEntity = ({
  columnData: { collection, headings },
  cellData,
  updateData,
}) => {
  const [editMode, setEditMode] = useState(false);
  return (
    <div className="dataTableCellData dataTableCellData-select">
      {editMode && (
        <PopupPanel isOpen={editMode} handleClose={() => setEditMode(false)}>
          <SelectEntityPanel
            currentSelection={cellData}
            collection={collection}
            headings={headings}
            handleSubmit={updateData}
          />
        </PopupPanel>
      )}
      {!editMode && (
        <button type="button" className="button-reset" onClick={() => setEditMode(true)}>
          {cellData}
        </button>
      )}
    </div>
  );
};

DataTableCellEntity.propTypes = {
  cellData: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  updateData: PropTypes.func.isRequired,
  columnData: PropTypes.shape({
    collection: PropTypes.string.isRequired,
    headings: PropTypes.objectOf(
      PropTypes.shape({
        uid: PropTypes.string.isRequired,
        label: PropTypes.string.isRequired,
        type: PropTypes.oneOf(Object.keys(filterTypes)).isRequired,
      })
    ),
  }).isRequired,
};

DataTableCellEntity.defaultProps = {
  cellData: 0,
};
