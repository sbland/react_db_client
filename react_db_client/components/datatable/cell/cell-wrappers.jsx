import React from 'react';
import PropTypes from 'prop-types';
import { Emoji } from '@react_db_client/components.emoji';
// import '../_dataTable.scss';
// import { RightClickWrapper } from '@samnbuk/react_db_client.components.popup-menu';
import { DataTableCellReadOnly } from '@samnbuk/react_db_client.components.datatable.cell-types';

/**
 * Data Table cell wrap that handles cell hovering
 *
 * @param {*} { className, style, handleHover, children }
 */
export const DataTableCellHoverWrap = ({
  className,
  style,
  handleHover,
  // columnWidth,
  disabled,
}) => (
  <div
    className={`dataTableCell_wrap ${className} ${disabled ? 'disabled' : 'enabled'}`}
    onFocus={() => handleHover && handleHover(true)}
    onMouseEnter={() => handleHover && handleHover(true)}
    onMouseLeave={() => handleHover && handleHover(false)}
    onBlur={() => handleHover && handleHover(false)}
    style={{
      // width: '100%',
      // height: '100%',
      ...style,
    }}
  />
);

DataTableCellHoverWrap.propTypes = {
  className: PropTypes.string,
  style: PropTypes.object,
  handleHover: PropTypes.func,
  columnWidth: PropTypes.number,
  disabled: PropTypes.bool,
};

DataTableCellHoverWrap.defaultProps = {
  className: '',
  style: {},
  handleHover: () => {},
  columnWidth: 100,
  disabled: false,
};

/**
 *
 * @param {*} param0
 * @returns
 */
export const EditColumnCell = ({
  allowSelection,
  allowRowDelete,
  allowRowEditPanel,
  isSelected,
  rowIndex,
  handleRemoveFromSelection,
  handleAddToSelection,
  handleDeleteRow,
  handleEditPanelBtnClick,
  rowUid,
}) => {
  return (
    <>
      {allowSelection && (
        <input
          className="rowSelectionBox"
          type="checkbox"
          id={rowIndex}
          checked={isSelected || false}
          onChange={() =>
            isSelected
              ? handleRemoveFromSelection(rowUid, rowIndex)
              : handleAddToSelection(rowUid, rowIndex)
          }
        />
      )}
      {allowRowDelete && (
        <button
          type="button"
          className="rowDeleteBtn"
          onClick={() => {
            handleDeleteRow(rowUid, rowIndex);
          }}
        >
          <Emoji emoj="🗑️" label="Delete" />
        </button>
      )}
      {allowRowEditPanel && (
        <button
          type="button"
          onClick={() => {
            handleEditPanelBtnClick(rowUid, rowIndex);
          }}
        >
          <Emoji emoj="📖" label="Open" />
        </button>
      )}
    </>
  );
};

EditColumnCell.propTypes = {
  allowSelection: PropTypes.bool.isRequired,
  allowRowDelete: PropTypes.bool.isRequired,
  allowRowEditPanel: PropTypes.bool.isRequired,
  isSelected: PropTypes.bool.isRequired,
  rowIndex: PropTypes.number.isRequired,
  handleRemoveFromSelection: PropTypes.func.isRequired,
  handleAddToSelection: PropTypes.func.isRequired,
  handleDeleteRow: PropTypes.func.isRequired,
  handleEditPanelBtnClick: PropTypes.func.isRequired,
  rowUid: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
};

// export const CellRightClickWrapper = ({ readOnly, clearCell, setAsDefault, children }) => {
//   if (readOnly) {
//     return children || '';
//   }

//   return (
//     <RightClickWrapper
//       items={[
//         { uid: 'clearCell', label: 'Clear', onClick: clearCell },
//         {
//           uid: 'setDefault',
//           label: 'Set as Default',
//           onClick: setAsDefault,
//         },
//       ]}
//       //TODO: How do we make this generic?
//       popupRoot="root"
//     >
//       {children}
//     </RightClickWrapper>
//   );
// };

// CellRightClickWrapper.propTypes = {
//   readOnly: PropTypes.bool,
//   clearCell: PropTypes.func.isRequired,
//   setAsDefault: PropTypes.func.isRequired,
//   children: PropTypes.oneOfType([PropTypes.node, PropTypes.arrayOf(PropTypes.nodes)]).isRequired,
// };

// CellRightClickWrapper.defaultProps = {
//   readOnly: false,
// };

const defaultComponent = () => (props) => <DataTableCellReadOnly {...props} />;

/**
 * Data Table Cell
 * Selects the correct cell type based on headings
 * Adds a right click wrapper
 *
 * @param {*} {
 *   rowId,
 *   cellData,
 *   columnData,
 *   updateData,
 * }
 * @returns
 */
// export const DataTableDataCell = (props) => {
//   const {
//     columnData: { type, readOnly, defaultValue },
//     updateData,
//     rowId,
//     columnId,
//     isDisabled,
//     componentMap,
//   } = props;
//   // TODO: Make sure read only component can parse any data
//   // const cellComponent = useMemo(
//   //   () =>
//   //     readOnly ? (
//   //       <DataTableCellReadOnly {...props} />
//   //     ) : (
//   //       switchF(type, componentMap, defaultComponent)
//   //     ),
//   //   [type, componentMap]
//   // );

//   const CellComponent = useMemo(
//     () => switchF(type, componentMap, defaultComponent),
//     [type, componentMap, defaultComponent]
//   );

//   // TODO: Can this not also handle read only?
//   if (isDisabled) {
//     return (
//       <div style={{ height: '100%' }} className="disabled">
//         <CellComponent {...props} />
//       </div>
//     );
//   }

//   return (
//     <div style={{ height: '100%' }}>
//         <CellComponent {...props} />
//     </div>
//   );
// };

// DataTableDataCell.propTypes = {
//   rowId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
//   columnId: PropTypes.string.isRequired,
//   rowData: PropTypes.shape({}).isRequired,
//   columnData: PropTypes.shape({
//     uid: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
//     label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
//     type: PropTypes.string.isRequired,
//     action: PropTypes.func,
//     to: PropTypes.string,
//     readOnly: PropTypes.bool,
//     defaultValue: PropTypes.any,
//   }).isRequired,
//   updateData: PropTypes.func.isRequired,
//   componentMap: PropTypes.objectOf(PropTypes.elementType).isRequired,
//   isDisabled: PropTypes.bool,
// };

// DataTableDataCell.defaultProps = {
//   isDisabled: false,
// }