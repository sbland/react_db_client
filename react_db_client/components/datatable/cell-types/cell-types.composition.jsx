import React, { useState } from 'react';

// eslint-disable-next-line import/no-extraneous-dependencies
import { DataTableCellText } from './DataTableCellText';
import { DataTableCellNumber } from './DataTableCellNumber';
import { DataTableCellEntity } from './DataTableCellEntity';
import { DataTableCellSelect } from './DataTableCellSelect';
import { filterTypes } from '@react_db_client/constants.client-types';

export const CellText = () => {
  const [data, setData] = useState('hello');
  return (
    <DataTableCellText
      columnData={{
        readOnly: false,
        type: 'text',
      }}
      cellData={data}
      updateData={setData}
      acceptValue={setData}
    />
  );
};

export const CellNumber = () => {
  const [data, setData] = useState('4.1234');
  return (
    <>
      <div>
        <DataTableCellNumber
          columnData={{
            defaultValue: 3.01,
            readOnly: false,
            min: 0,
            max: 30.3,
            step: 0.1,
            type: 'number',
          }}
          cellData={data}
          updateData={setData}
          acceptValue={setData}
          rowData={{}}
        />
      </div>
      <div>Cell data: {data}</div>
    </>
  );
};

export const CellSelect = () => {
  const [data, setData] = useState('a');
  return (
    <>
      <div>
        <DataTableCellSelect
          columnData={{
            type: 'select',
            options: [
              { uid: 'a', label: 'A' },
              { uid: 'b', label: 'B' },
              { uid: 'c', label: 'C' },
            ],
          }}
          cellData={data}
          updateData={setData}
          acceptValue={setData}
          rowData={{}}
        />
      </div>
      <div>Cell data: {data}</div>
    </>
  );
};

export const CellEntity = () => (
  <DataTableCellEntity
    columnData={{
      collection: 'products',
      headings: {
        a: { uid: 'a', label: 'A', type: filterTypes.text },
        b: { uid: 'b', label: 'B', type: filterTypes.number },
      },
    }}
    cellData="Foo"
    updateData={console.log}
  />
);
