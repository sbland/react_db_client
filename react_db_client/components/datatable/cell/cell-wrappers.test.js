import '@samnbuk/react_db_client.testing.enzyme-setup';
import React from 'react';
import { mount } from 'enzyme';
import { DataTableCellText } from '@samnbuk/react_db_client.components.datatable.cell-types';
import { RightClickWrapper } from '@samnbuk/react_db_client.components.popup-menu';
import { DataTableDataCell } from './cell-wrappers';


describe('DataTableDataCell', () => {
  const defaultProps = {
    columnData: { type: 'text', readOnly: false, def: 'b', uid: 'a', label: 'a' },
    updateData: () => {},
    rowId: 'row1',
    columnId: 'col1',
    cellData: 'a',
    rowData: {
      uid: 'a',
      col1: 'a',
    },
    componentMap: {},
  };
  it('Renders', () => {
    mount(<DataTableDataCell {...defaultProps} />);
  });
  it('Matches Snapshot', () => {
    const component = mount(<DataTableDataCell {...defaultProps} />);
    const tree = component.debug();
    expect(tree).toMatchSnapshot();
  });
  describe('Unit Tests', () => {
    describe('Basic', () => {
      const component = mount(<DataTableDataCell {...defaultProps} />);
      it('Renders a text cell', () => {
        expect(component.find('.dataTableCellData')).toBeTruthy();
        expect(component.find(RightClickWrapper)).toBeTruthy();
        expect(component.find(DataTableCellText)).toBeTruthy();
      });
    });
    describe('Using custom field components', () => {
      test.skip('should allow passing custom field component', () => {
        //
      });
    });
  });
});
