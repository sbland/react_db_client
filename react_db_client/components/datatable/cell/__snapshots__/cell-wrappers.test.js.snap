// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableDataCell Matches Snapshot 1`] = `
"<DataTableDataCell columnData={{...}} updateData={[Function: updateData]} rowId=\\"row1\\" columnId=\\"col1\\" cellData=\\"a\\" rowData={{...}} componentMap={{...}} isDisabled={false}>
  <div style={{...}}>
    <CellRightClickWrapper readOnly={false} clearCell={[Function: clearCell]} setAsDefault={[Function: setAsDefault]}>
      <RightClickWrapper items={{...}} popupRoot=\\"root\\">
        <div style={{...}} className=\\"popupMenu_righClickWrap false\\" onContextMenu={[Function: handleContextMenu]}>
          <Component columnData={{...}} updateData={[Function: updateData]} rowId=\\"row1\\" columnId=\\"col1\\" cellData=\\"a\\" rowData={{...}} componentMap={{...}} isDisabled={false}>
            <DataTableCellReadOnly columnData={{...}} updateData={[Function: updateData]} rowId=\\"row1\\" columnId=\\"col1\\" cellData=\\"a\\" rowData={{...}} componentMap={{...}} isDisabled={false}>
              <div className=\\"dataTableCellData readOnly\\">
                <div type=\\"button\\" className=\\"dataTableCellData_text\\">
                  a
                </div>
              </div>
            </DataTableCellReadOnly>
          </Component>
        </div>
        <PopupMenu isOpenOverride={false} onCloseCallback={[Function: onCloseCallback]} items={{...}} position={{...}} popupRoot=\\"root\\">
          <div>
            Missing Popup Root
          </div>
        </PopupMenu>
      </RightClickWrapper>
    </CellRightClickWrapper>
  </div>
</DataTableDataCell>"
`;
